Source/ULibs.TinyJsonDeser/JsonDeserializer.cs

```diff
@@ -144,7 +145,7 @@ public bool TryParseString(string json, out string output)
         /// <param name="json">The json string to parse.</param>
         /// <param name="output">If the parse succeeded, this will hold the parsed result.</param>
         /// <returns>Whether or not the parse attempt succeeded.</returns>
-        public bool TryParseArray(string json, out object[] output)
+        public bool TryParseArray(string json, out object?[]? output)

```

Comment:
Shouldn't this method, and all of the other TryParse methods, also have the `[NotNullWhen(true)]` attribute?

Reply:
So I initially tried that but it turns out it is possible for `TryParseValue` to set `output = null` when returning `true` (see line 199). This means its callers also can't be sure, though really this only matters to its overload.That said, I have found a couple of places where that wasn't an issue and have added them

Reply:
Oh, that's worrying. As far as I can remember, all of the TryParse methods should produce a non-null result if the parse succeeds, with a mere handful of notable exceptions (e.g. TryParseValue, where `null` is a legitimate JSON value). If that's not true, there might be bugs.

Reply:
I'm struggling to remember, though. It's a long time since I wrote this.

Reply:
I will take a closer look at the TryParseNull method

Reply:
I should probably checkout the code and have a proper look, rather than just commenting on the changes.

Reply:
It occurs to me that we could avoid the need for the NullabilityAttribute hack altogether by returning non-null sentinel values for all of the parse methods for reference types. Something like this:

```c#
public static bool TryParseString(string json, ref int offset, out string value)
{
    if (<can-parse>)
    {
        value = <parsed-value>;
        return true;
    }
    value = "";  / Sentinel value to ensure we always output a non-null value.
    return false;
}

public static bool TryParseArray(string json, ref int offset, out object?[] value)
{
    if (<can-parse>)
    {
        value = <parsed-value>;
        return true;
    }
    value = new object[0];  // Sentinel value to ensure we always output a non-null value.
    return false;
}
```

What do you think?

Reply:
That feels like it would be nicer - its already done in a few of the parsing methods

Reply:
I've tried the above on a separate branch. Here's the PR, so you can see the changes vs the original code on master. I'm not poposing it be merged. It's just an illustration of an alternative approach. If we're happy with it, we can merge that into this branch. If not, delete it.

Reply:
To me it feels more correct/ergonomic to return null when a `TryParse` function fails, since it's a genuine lack of a value and users might accidentally use an incorrect value otherwise?

Reply:
Shall I bring that change into this branch, then? I might have to have a look at any changes neccessary in the tests, first.

Reply:
Actually, that's nonsense. The tests don't care about nullability. Maybe they should?

Reply:
Yeah, more tests around null values would probably be good - I think we hit a few cases earlier where we changed null behavior but the tests didn't break :) John added a few extra test cases, but more is always good...

Reply:
I meant enabling null reference types in the test project.It's currently at C# 7.3 with null reference types disabled.

Reply:
I think I had ignored it because changing the language meant changing the target framework and the test projects have old-style csproj and it all felt like too much work for debatable benefit - however I may have misunderstood what was needed

Reply:
Okay, I've merged the change that uses non-null default values for when the TryParse methods fail.

Reply:
> To me it feels more correct/ergonomic to return null when a TryParse function fails, since it's a genuine lack of a value and users might accidentally use an incorrect value otherwise?

Sorry. I missed this, I think in a mid-air collision. I only just saw it when I hit refresh.

I'm not sure I agree that it's more correct or ergonomic to return null. What the contract really wants to express is that the output value is undefined when the parse fails, but the language and type system isn't rich enough to really support that yet and we have to set the output parameter to something. But there's nothing special about null here. Returning null would still be defined behaviour (we'd be defining it), and users might still accidentally try to use it without checking the return value first.

The emerging fix to this scenario is the `NotNullWhenAttribute` (the real one, not our fake), but it's still a kludge. It doesn't truly give us type safety in the language. It doesn't even give us a compilation error by default, either. Just a warning.

It'd be nice if there was a ubiquitous `Option<T>` type we could rely on, so callers genuinely only get access to the value if the parse succeeds, but sadly there isn't yet. And C# devs tend to find that sort of thing fugly, still, though it's been getting progressively better in recent years.

One final potential problem with taking the fake attribute approach. We'd be defining a `System.Diagnostics.CodeAnalysis.NotNullWhenAttribute` type in the consuming project. If the consumer tries to make use of two or more libraries that do this, the names would conflict, and boom! It's not a sustainable pattern for us to follow.

As I've typed this comment, I've been somewhat on the fence, but that last issue is a clincher for me. I think it's asking for trouble to declare a polyfill for a system type and expose it to users like that. I think it's better to accept the risk that users might mess up when trying to use a library that follows a fairly idiomatic pattern in .NET. Too bad for them if they get it wrong, but I think the risk is low _(and egg on their faces, not ours, if they do go wrong!)_.

Reply:
> _(and egg on their faces, not ours, if they do go wrong!)_.

I don't think that's a particularly good attitude for library development - we should all be about the "pit of success" here 😛 

I totally agree with the rest of your arguments, though - I guess if we wanted to do the attribute polyfill properly we'd need a whole other microlibrary which consumers could independently import?

Reply:
> I don't think that's a particularly good attitude for library development - we should all be about the "pit of success" here 😛

Sorry, I was kinda joking. More seriously, I agree, but the only real options for making this any better either involve a significantly breaking API change (e.g return an option type) or the attribute hack that's potentially dangerous. And TryParse is such a common idiom in .NET that the current code feels pretty reasonable and low risk. I've been exploring ways to have multiple versions of this file, so that where the `WhereNotNull` is available, we use it, but when  it's not we fall back to the current implementation. I managed to get it working, but I think it's best to leave it for another PR.

> I totally agree with the rest of your arguments, though - I guess if we wanted to do the attribute polyfill properly we'd need a whole other microlibrary which consumers could independently import?

Yeah. We could possibly do that as a conditional dependency. I'll have a play with that, too.